### DISCUSSION TITLE
希望排除程序能添加个白名单模式

### DISCUSSION THREAD

**Discussion by OkinoAze** [shuax/MouseInc.Settings#195] [Ideas]
电脑上大部分软件使用时用不到手势的功能,但一个个排除有点花时间,希望能做个白名单的模式指定需要的软件使用手势.
(*已经使用WGestures 2替代,解决了所有问题)

**Comment by 6Tom** [comment]
我不同意这个提议，只会给耍下大佬增加无用工作量，下面我阐述理由。

其他软件如果你**不想用手势，不要画手势就是了**，为什么要排除。如果需要调用的全局热键与软件热键冲突，可以**临时暂停**mouseinc。

**排除的作用是**，有些软件它需要右键滑动，不排除的话就得来回暂停、启用、暂停、启用，这不太方便。**为了避免来回切换**，所以有了排除功能。

需要右键滑动的软件无非就是少数**建模、渲染、三维贴图软件**，你不可能满电脑的软件都需要排除。

再这么好用的手势，为什么**“大部分软件使用时用不到手势的功能”**呢？**最大化、还原、最小化、alt拖动**都用不到？**我非常不理解**。

**Reply by OkinoAze** [reply]
从提议来看我认为这是可行的,不过好像作者不怎么关注.
实际白名单就是黑名单取反,实质并不增加多少工作量.像很多软件写规则也有白名单和黑名单模式.
举个例子可能会明白,如果有玩家玩游戏的时候都是以窗口化运行,先不谈软件拖动或者点击会出现帧数波动的问题.
大部分玩家拥有不下百款游戏,而大部分游戏都有右键指定的操作方式,这时就必须暂停软件的功能或者排除.作为用户来说暂时开关和排除都需要一个步骤,而累积起来就比较麻烦.
除此之外我自身就使用非常多的软件,比如画图做动画,3d还有音乐加上一些杂七杂八的小工具.这里面有些本身就有功能上的冲突.
而右键拖拽突然就会小卡一下和出现手势图案都是比较影响使用体验.我自己的黑名单列表就有不下百款软件.
不过我的使用场景就单一个浏览器控制,所以有考虑过再换回浏览器插件的情况.

**Reply by 6Tom** [reply]
我猜你用的是奔跑的奶酪定制款浏览器吧，因为奶子传递给大家的思想基本就是：mouseinc是可以拿来增强非国产浏览器的，全局手势软件。这我不认同，mouseinc根本就不是为浏览器定制的，这是完完全全的大材小用。

我不玩任何游戏，游戏我确实一窍不通。但是游戏是不是还可以用“全屏排除”？我猜大部分游戏是全屏的？或者说相当一部分游戏是全屏？或者说很多游戏具备全屏的功能？

但是我专业建模，所以我知道哪些环境下右键滑动冲突。而且并非所有建模软件右键都冲突，右键不冲突的软件甚至可以结合mouseinc增强特殊功能。

实话实说，你需要排除上百款软件，以我的认知，我是不相信的。根本不会有那么多软件冲突，我电脑上常用86款软件，再算上不常用，总计超过百款软件。但是只有两款需要排除，分别是lumion和rhino，另有photoshop、indesign也冲突，但是这两款我宁可临时暂停，因为手势和拖拽我仍然有时需要，他太方便了。

**Comment by OkinoAze** [comment]
我自身有七百余款游戏,因为显示器分辨率高和太大所以运行的时候会使用窗口模式游玩.
其中一些游戏在窗口模式下体验会更好.

并且像是工具类的软件 比如图片查看器 压缩软件 音乐播放软件 视频播放软件 等基本上默认运行是在窗口模式下,一般这类软件使用场景很简单,所以一般也用不到手势的操作.

而关键是在这款工具滑动的时候会造成遮挡和卡顿,对于游戏玩家拖动卡顿和右键操作上的冲突是不能忍受的.
就算在通用设置里不设置任何快捷操作,依旧会有轨迹和文字提示.所以必须添加到排除列表里.

当然这款软件对于我来说是方便的,在浏览器插件里必须要载完网页 一些设置页面不浏览器插件也不能使用手势能操作,
这款软件在插件局限的地方表现都很好.

但对于类似与我这种就单一使用在浏览器上或者零星的几个软件上的用户群体来说,每次添加到排除列表总归是麻烦的一件事情.
而且你阐述的观点也不能解决有这方面需求的用户的问题,再者添加白名单类似的功能并不需要很花时间,我觉得作者是可以考虑的.

**Reply by 6Tom** [reply]
既然我无法说服你，或者让你临时暂停使用mouseinc，那我就换个思路帮你把问题解决了吧。

帮你做了个批处理，示例是一键获取d盘所有exe文件的名称，加引号、加逗号、加空格，生成的结果在批处理同一目录下的Excludes.txt中。最终结果会只有一行特别长的行，全选复制，到MouseInc.json中，查找到“Excludes”字符串，把复制的内容粘贴到它后面的中括号里，就实现了排除d盘所有exe。

如你所言，你排除的exe会比不排除的exe多，那你可以在生成的结果中，自行查找不需要排除的几个exe，删除它，就算是实现了“白名单”。我的d盘有686个exe文件，测试用时大约4.5秒，文件越多越慢，等几秒就好了。如果需要换其他盘符，就自行改一下变量foo的值吧，然后自己合并一下结果，注意合并两个结果之间自行加逗号和空格。

代码如下，请保存为utf8格式的bat。

```
@echo off
chcp 65001
set "foo=d:\"
cd /d "%~dp0"
echo/Wait a few seconds.
(for /f "delims=" %%a in ('dir /s /a-d /b "%foo%"^|find /i ".exe"') do if "%%~xa"==".exe" echo/%%~nxa)>"temp1.txt"
for /f "tokens=1* delims=:" %%a in ('findstr /n ".*" "temp1.txt"') do if %%a==1 echo/"%%b">"temp2.txt"
(for /f "tokens=1* delims=:" %%a in ('findstr /n ".*" "temp1.txt"') do if %%a neq 1 echo/, "%%b")>>"temp2.txt"
(for /f "usebackq delims=" %%a in ("temp2.txt") do set /p ="%%a"<nul)>"Excludes.txt"
del /f /q "temp*.txt"
```

**Reply by raccoon666666** [reply]
他可能不知道批处理怎么用

**Reply by 6Tom** [reply]
不知道，那就看他自己了

**Reply by OkinoAze** [reply]
看样子还是没理解重点,就算使用bat获取到了exe,多出需要等待的时间不说,如果新增了不需要的软件,还是需要额外的一系列操作.
从结果上说也还是手动操作,这和手动开关,画一个手势添加到黑名单没有太大的区别.
而且如果你把自己设为一位普通的用户来说,就像你所说"那就看他自己了一样",
那么这之中普通用户多出来要找到你贴的bat时间,还有等待时间,还有如上面一位所说"他可能不知道批处理怎么用",这些也是需要考量的.
可能对于白名单的功能请求或许只有我一个人需要使用到,但不论如何对于我自身来说添加白名单功能比弯弯绕绕的解决办法来的更直接.

**Reply by 6Tom** [reply]
等待一会，也比你一个一个排除要省事多了吧。

你要是有新游戏入账了，就从那之后，每添加一个新游戏排除一个吧。

因为就如你自己所说，这个需求可能只有你一个人需要，咱们不了解shuax大佬的意愿，或者即便愿意，技术难度咋样，时间精力成本咋样。

在作者接受你的建议之前，暂时先这样解决吧。

**Comment by shuax** [comment]
https://github.com/shuax/MouseInc.Settings/discussions/234#discussioncomment-2170915
